File: 073_TopKFrequentWords.cpp
// Topic: Top K Frequent Words
// Description: Given a list of words, return the k most frequent words. If frequencies tie, return
// lexicographically smaller first. Approach: Count frequencies with unordered_map, then use vector
// + sort with custom comparator (or min-heap). Here we sort the unique words by freq desc, then
// lexicographically.

#include <bits/stdc++.h>
using namespace std;

vector<string> topKFrequent(vector<string>& words, int k)
{
    unordered_map<string, int> cnt;
    for (auto& w : words) cnt[w]++;
    vector<string> uniq;
    uniq.reserve(cnt.size());
    for (auto& p : cnt) uniq.push_back(p.first);
    sort(uniq.begin(), uniq.end(),
         [&](const string& a, const string& b)
         {
             if (cnt[a] != cnt[b])
                 return cnt[a] > cnt[b];
             return a < b;
         });
    if ((int)uniq.size() > k)
        uniq.resize(k);
    return uniq;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<string> words = {"i", "love", "leetcode", "i", "love", "coding"};
    auto ans = topKFrequent(words, 2);
    for (auto& s : ans) cout << s << "\n";
}

/*
Time Complexity: O(m log m) where m = number of unique words (plus O(n) to count)
Space Complexity: O(m)
*/

File: 073_TopKFrequentWords.py
# Topic: Top K Frequent Words
# Description: Given a list of words, return the k most frequent words. If frequencies tie, return lexicographically smaller first.
# Approach: Count frequencies with Counter, then sort with custom comparator.

from typing import List
from collections import Counter


class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        cnt = Counter(words)
        # Sort by frequency desc, then alphabetically asc
        uniq = sorted(cnt.keys(), key=lambda w: (-cnt[w], w))
        return uniq[:k]


if __name__ == "__main__":
    words = ["i", "love", "leetcode", "i", "love", "coding"]
    result = Solution().topKFrequent(words, 2)
    for s in result:
        print(s)

"""
Time Complexity: O(m log m) where m = number of unique words (plus O(n) to count)
Space Complexity: O(m)
"""
